File: c/tests/test_vision_cloud.c
/*
 * test_vision_cloud.c — CircleAI.Vision.Cloud IImageGenerator (C11 port).
 *
 * Verifies NullImageGenerator, the deterministic fake generator (configured vs
 * not, Count clamp 1..4, deterministic clock), and the ImageGeneratorFallbackChain
 * (skip unconfigured, first non-empty wins, DisplayLabel/StatusMessage).
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "circle_ai/circle_ai.h"

static void test_null_generator(void) {
    ca_image_generator_t g = ca_null_image_generator();
    assert(strcmp(ca_image_generator_id(&g), "null") == 0);
    assert(strcmp(ca_image_generator_display_label(&g), "No image generator") == 0);
    assert(!ca_image_generator_is_configured(&g));
    assert(strstr(ca_image_generator_status_message(&g), "OpenAI:ApiKey") != NULL);

    ca_image_generation_request_t req;
    ca_image_generation_request_init(&req, "a cat");
    assert(req.size == 1024 && req.count == 1);
    size_t n = 99;
    ca_image_artifact_t *arr = ca_image_generator_generate(&g, &req, &n);
    assert(arr == NULL && n == 0);
    printf("  null_generator: ok\n");
}

static void test_fake_generator(void) {
    /* Unconfigured -> empty. */
    ca_fake_image_generator_t *fu = ca_fake_image_generator_create("openai-images", "OpenAI", false, 5000);
    ca_image_generator_t gu = ca_fake_image_generator_as_iface(fu);
    assert(!ca_image_generator_is_configured(&gu));
    ca_image_generation_request_t req;
    ca_image_generation_request_init(&req, "sunset");
    size_t n = 7;
    ca_image_artifact_t *arr = ca_image_generator_generate(&gu, &req, &n);
    assert(arr == NULL && n == 0);
    ca_fake_image_generator_destroy(fu);

    /* Configured -> Count clamped to 1..4, deterministic url + clock. */
    ca_fake_image_generator_t *fc = ca_fake_image_generator_create("stability", "Stability", true, 5000);
    ca_image_generator_t gc = ca_fake_image_generator_as_iface(fc);
    assert(ca_image_generator_is_configured(&gc));
    assert(strcmp(ca_image_generator_id(&gc), "stability") == 0);
    assert(strcmp(ca_image_generator_display_label(&gc), "Stability") == 0);

    req.count = 9;   /* clamps to 4 */
    arr = ca_image_generator_generate(&gc, &req, &n);
    assert(n == 4 && arr);
    for (size_t i = 0; i < n; ++i) {
        assert(strcmp(arr[i].generator_id, "stability") == 0);
        assert(strcmp(arr[i].prompt, "sunset") == 0);
        assert(strcmp(arr[i].mime_type, "image/png") == 0);
        assert(arr[i].url && strncmp(arr[i].url, "mem://stability/sunset/", 23) == 0);
        assert(arr[i].bytes == NULL);
        assert(arr[i].generated_at_utc_ms == 5000);
    }
    ca_image_artifact_free_array(arr, n);

    req.count = 0;   /* clamps to 1 */
    arr = ca_image_generator_generate(&gc, &req, &n);
    assert(n == 1 && arr);
    ca_image_artifact_free_array(arr, n);
    ca_fake_image_generator_destroy(fc);
    printf("  fake_generator: ok\n");
}

static void test_fallback_chain(void) {
    /* Chain: [unconfigured-A, configured-B, configured-C]. B serves first. */
    ca_fake_image_generator_t *a = ca_fake_image_generator_create("a", "A", false, 1);
    ca_fake_image_generator_t *b = ca_fake_image_generator_create("b", "B", true, 2);
    ca_fake_image_generator_t *c = ca_fake_image_generator_create("c", "C", true, 3);
    ca_image_generator_t members[3] = {
        ca_fake_image_generator_as_iface(a),
        ca_fake_image_generator_as_iface(b),
        ca_fake_image_generator_as_iface(c),
    };
    ca_image_generator_fallback_chain_t *chain =
        ca_image_generator_fallback_chain_create(members, 3, false);
    assert(ca_image_generator_fallback_chain_count(chain) == 3);
    ca_image_generator_t cg = ca_image_generator_fallback_chain_as_iface(chain);

    assert(strcmp(ca_image_generator_id(&cg), "fallback-chain") == 0);
    assert(strcmp(ca_image_generator_display_label(&cg), "Fallback (3)") == 0);
    assert(ca_image_generator_is_configured(&cg));
    /* StatusMessage lists configured ids in order: "Ready · b → c". */
    const char *status = ca_image_generator_status_message(&cg);
    assert(strstr(status, "Ready") != NULL);
    assert(strstr(status, "b") != NULL && strstr(status, "c") != NULL);

    ca_image_generation_request_t req;
    ca_image_generation_request_init(&req, "moon");
    size_t n = 0;
    ca_image_artifact_t *arr = ca_image_generator_generate(&cg, &req, &n);
    assert(n == 1 && arr);
    /* First non-empty came from B (2), NOT the unconfigured A nor C. */
    assert(strcmp(arr[0].generator_id, "b") == 0);
    assert(arr[0].generated_at_utc_ms == 2);
    ca_image_artifact_free_array(arr, n);

    ca_image_generator_fallback_chain_destroy(chain);
    ca_fake_image_generator_destroy(a);
    ca_fake_image_generator_destroy(b);
    ca_fake_image_generator_destroy(c);

    /* All-unconfigured chain -> empty result + "No configured generator" status. */
    ca_fake_image_generator_t *u1 = ca_fake_image_generator_create("u1", "U1", false, 1);
    ca_image_generator_t only = ca_fake_image_generator_as_iface(u1);
    ca_image_generator_fallback_chain_t *empty =
        ca_image_generator_fallback_chain_create(&only, 1, false);
    ca_image_generator_t eg = ca_image_generator_fallback_chain_as_iface(empty);
    assert(!ca_image_generator_is_configured(&eg));
    assert(strcmp(ca_image_generator_status_message(&eg), "No configured generator in chain.") == 0);
    arr = ca_image_generator_generate(&eg, &req, &n);
    assert(arr == NULL && n == 0);
    ca_image_generator_fallback_chain_destroy(empty);
    ca_fake_image_generator_destroy(u1);
    printf("  fallback_chain: ok\n");
}

/* Chain with own=true must destroy its members on destroy (leak-checked by ASan
 * in CI; here we just confirm it runs cleanly with heap-backed fakes). */
static void test_fallback_chain_owning(void) {
    ca_fake_image_generator_t *b = ca_fake_image_generator_create("b", "B", true, 2);
    /* Wrap the fake in an owning iface by supplying a destroy thunk. */
    ca_image_generator_t iface = ca_fake_image_generator_as_iface(b);
    /* as_iface leaves destroy NULL; emulate an owning member by setting it. */
    iface.destroy = (void (*)(void *))ca_fake_image_generator_destroy;
    iface.self = b;
    ca_image_generator_fallback_chain_t *chain =
        ca_image_generator_fallback_chain_create(&iface, 1, true);
    /* destroy the chain -> it calls destroy(b). */
    ca_image_generator_fallback_chain_destroy(chain);
    printf("  fallback_chain_owning: ok\n");
}

int main(void) {
    test_null_generator();
    test_fake_generator();
    test_fallback_chain();
    test_fallback_chain_owning();
    printf("test_vision_cloud: all assertions passed\n");
    return 0;
}

File: python/tests/test_vision_cloud.py
"""test_vision_cloud.py — CircleAI.Vision.Cloud image-generation surface.

Covers the null generator, the OpenAI + Stability generators over the injected
in-memory HTTP transport (exact request construction + response parsing), and the
fallback-chain composite (skip-unconfigured, first-non-empty). C#
(CircleAI.Vision.Cloud) is the reference.
"""
from __future__ import annotations

import json

import pytest

from circle_ai.vision.cloud import (
    HttpResponse,
    IImageGenerator,
    ImageGenerationRequest,
    ImageGeneratorFallbackChain,
    InMemoryImageHttpTransport,
    NullImageGenerator,
    OpenAiImageGenerator,
    OpenAiImageOptions,
    StabilityImageGenerator,
    StabilityImageOptions,
)


# ── null generator ───────────────────────────────────────────────────────────────


async def test_null_image_generator_returns_empty():
    g = NullImageGenerator.instance()
    assert g.generator_id == "null"
    assert g.display_label == "No image generator"
    assert g.is_configured is False
    assert "Configure OpenAI:ApiKey" in g.status_message
    assert await g.generate_async(ImageGenerationRequest(prompt="a cat")) == ()


# ── OpenAI generator ─────────────────────────────────────────────────────────────


def _openai_ok(urls):
    body = json.dumps({"data": [{"url": u} for u in urls]})
    return lambda req: HttpResponse(status_code=200, body_text=body)


async def test_openai_not_configured_short_circuits():
    transport = InMemoryImageHttpTransport(_openai_ok(["http://x/1.png"]))
    g = OpenAiImageGenerator(transport, OpenAiImageOptions(api_key=None))
    assert g.is_configured is False
    assert "OpenAI API key not configured" in g.status_message
    assert await g.generate_async(ImageGenerationRequest(prompt="cat")) == ()
    assert transport.requests == []  # never hits the wire when unconfigured


async def test_openai_builds_exact_request_and_parses_urls():
    transport = InMemoryImageHttpTransport(_openai_ok(["http://cdn/a.png", "http://cdn/b.png"]))
    g = OpenAiImageGenerator(transport, OpenAiImageOptions(api_key="sk-test", model="dall-e-3"))
    assert g.generator_id == "openai-images"
    assert g.display_label == "OpenAI · dall-e-3"
    assert g.is_configured is True

    arts = await g.generate_async(ImageGenerationRequest(prompt="a red bus", size=512, count=2))
    # request shape
    req = transport.requests[0]
    assert req.method == "POST"
    assert req.path == "/v1/images/generations"
    assert req.headers["Authorization"] == "Bearer sk-test"
    assert req.json_body == {
        "model": "dall-e-3",
        "prompt": "a red bus",
        "n": 2,
        "size": "512x512",
        "response_format": "url",
    }
    # parsed artifacts
    assert len(arts) == 2
    assert arts[0].url == "http://cdn/a.png"
    assert arts[0].bytes_ is None
    assert arts[0].mime_type == "image/png"
    assert arts[0].generator_id == "openai-images"
    assert arts[0].prompt == "a red bus"


async def test_openai_clamps_count_to_four():
    transport = InMemoryImageHttpTransport(_openai_ok([]))
    g = OpenAiImageGenerator(transport, OpenAiImageOptions(api_key="sk"))
    await g.generate_async(ImageGenerationRequest(prompt="p", count=99))
    assert transport.requests[0].json_body["n"] == 4
    await g.generate_async(ImageGenerationRequest(prompt="p", count=0))
    assert transport.requests[1].json_body["n"] == 1


async def test_openai_non_success_returns_empty():
    transport = InMemoryImageHttpTransport(lambda req: HttpResponse(status_code=429, body_text="rate limited"))
    g = OpenAiImageGenerator(transport, OpenAiImageOptions(api_key="sk"))
    assert await g.generate_async(ImageGenerationRequest(prompt="p")) == ()


async def test_openai_skips_items_without_url():
    body = json.dumps({"data": [{"revised_prompt": "no url here"}, {"url": "http://ok/1.png"}]})
    transport = InMemoryImageHttpTransport(lambda req: HttpResponse(200, body_text=body))
    g = OpenAiImageGenerator(transport, OpenAiImageOptions(api_key="sk"))
    arts = await g.generate_async(ImageGenerationRequest(prompt="p"))
    assert len(arts) == 1
    assert arts[0].url == "http://ok/1.png"


# ── Stability generator ──────────────────────────────────────────────────────────


async def test_stability_not_configured_short_circuits():
    transport = InMemoryImageHttpTransport(lambda req: HttpResponse(200, body_bytes=b"\x89PNG"))
    g = StabilityImageGenerator(transport, StabilityImageOptions(api_key=""))
    assert g.is_configured is False
    assert "Stability AI API key not configured" in g.status_message
    assert await g.generate_async(ImageGenerationRequest(prompt="p")) == ()
    assert transport.requests == []


async def test_stability_loops_per_image_and_returns_bytes():
    transport = InMemoryImageHttpTransport(lambda req: HttpResponse(200, body_bytes=b"IMGDATA"))
    g = StabilityImageGenerator(transport, StabilityImageOptions(api_key="sk", model="sd3.5-large"))
    assert g.generator_id == "stability"
    assert g.display_label == "Stability AI · sd3.5-large"

    arts = await g.generate_async(ImageGenerationRequest(prompt="a dog", count=3))
    assert len(transport.requests) == 3  # one call per requested image
    assert len(arts) == 3
    a = arts[0]
    assert a.url is None
    assert a.bytes_ == b"IMGDATA"
    assert a.mime_type == "image/png"
    # request shape
    req = transport.requests[0]
    assert req.method == "POST"
    assert req.path == "/v2beta/stable-image/generate/sd3"
    assert req.headers["Authorization"] == "Bearer sk"
    assert req.headers["Accept"] == "image/png"
    assert req.form_fields == (("prompt", "a dog"), ("output_format", "png"), ("model", "sd3.5-large"))


async def test_stability_adds_negative_prompt_when_present():
    transport = InMemoryImageHttpTransport(lambda req: HttpResponse(200, body_bytes=b"X"))
    g = StabilityImageGenerator(transport, StabilityImageOptions(api_key="sk"))
    await g.generate_async(ImageGenerationRequest(prompt="p", negative_prompt="blurry", count=1))
    fields = dict(transport.requests[0].form_fields)
    assert fields["negative_prompt"] == "blurry"


async def test_stability_skips_failed_images_but_keeps_successes():
    calls = {"n": 0}

    def handler(req):
        calls["n"] += 1
        # first image fails, rest succeed
        if calls["n"] == 1:
            return HttpResponse(500, body_text="server error")
        return HttpResponse(200, body_bytes=b"OK")

    transport = InMemoryImageHttpTransport(handler)
    g = StabilityImageGenerator(transport, StabilityImageOptions(api_key="sk"))
    arts = await g.generate_async(ImageGenerationRequest(prompt="p", count=3))
    assert len(transport.requests) == 3
    assert len(arts) == 2  # only the two successes


async def test_stability_honours_output_format():
    transport = InMemoryImageHttpTransport(lambda req: HttpResponse(200, body_bytes=b"J"))
    g = StabilityImageGenerator(transport, StabilityImageOptions(api_key="sk", output_format="jpeg"))
    arts = await g.generate_async(ImageGenerationRequest(prompt="p", count=1))
    assert arts[0].mime_type == "image/jpeg"
    assert transport.requests[0].headers["Accept"] == "image/jpeg"


# ── fallback chain ───────────────────────────────────────────────────────────────


class _StubGen(IImageGenerator):
    def __init__(self, gid, configured, result):
        self._gid = gid
        self._configured = configured
        self._result = result
        self.called = False

    @property
    def generator_id(self):
        return self._gid

    @property
    def display_label(self):
        return self._gid

    @property
    def is_configured(self):
        return self._configured

    @property
    def status_message(self):
        return "ok"

    async def generate_async(self, request, ct=None):
        self.called = True
        return self._result


def _artifact_tuple():
    from datetime import datetime, timezone

    from circle_ai.vision.cloud import ImageArtifact

    return (ImageArtifact("g", "p", "image/png", "http://x/1.png", None, datetime.now(timezone.utc)),)


async def test_fallback_chain_empty_is_unconfigured():
    chain = ImageGeneratorFallbackChain(None)
    assert chain.generator_id == "fallback-chain"
    assert chain.display_label == "Fallback (0)"
    assert chain.is_configured is False
    assert chain.status_message == "No configured generator in chain."
    assert await chain.generate_async(ImageGenerationRequest(prompt="p")) == ()


async def test_fallback_chain_skips_unconfigured_and_takes_first_nonempty():
    unconfigured = _StubGen("a", configured=False, result=_artifact_tuple())
    configured_empty = _StubGen("b", configured=True, result=())
    configured_hit = _StubGen("c", configured=True, result=_artifact_tuple())
    late = _StubGen("d", configured=True, result=_artifact_tuple())

    chain = ImageGeneratorFallbackChain([unconfigured, configured_empty, configured_hit, late])
    assert chain.is_configured is True
    assert chain.display_label == "Fallback (4)"
    assert chain.status_message == "Ready · b → c → d"

    res = await chain.generate_async(ImageGenerationRequest(prompt="p"))
    assert len(res) == 1
    assert unconfigured.called is False  # skipped (not configured)
    assert configured_empty.called is True  # tried, returned empty
    assert configured_hit.called is True  # tried, returned the hit
    assert late.called is False  # never reached — first non-empty wins


async def test_fallback_chain_all_empty_returns_empty():
    a = _StubGen("a", True, ())
    b = _StubGen("b", True, ())
    chain = ImageGeneratorFallbackChain([a, b])
    assert await chain.generate_async(ImageGenerationRequest(prompt="p")) == ()
    assert a.called and b.called
